compiler/src/iree/compiler/Codegen/Common/ConvertUnsupportedFloatArithPass.cpp

```diff
@@ -0,0 +1,107 @@
+#include "iree/compiler/Codegen/Common/Transforms.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Dialect/AMDGPU/Utils/Chipset.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_CONVERTUNSUPPORTEDFLOATARITHPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+
+struct ConvertUnsupportedFloatArithPass final
+    : public impl::ConvertUnsupportedFloatArithPassBase<
+          ConvertUnsupportedFloatArithPass> {
+  void runOnOperation() override;
+  using Base::Base;
+};
+
+} // namespace
+//
+
+// Populates source and target conversion types based on the target
+// architecture.
+// TODO(pashu123): Refine the patterns based on the target arch.
+static LogicalResult
+populateSourceAndTargetType(MLIRContext *ctx, Operation *op,
+                            SmallVectorImpl<Type> &sourceTypes,
+                            Type &targetType) {
+
+  auto gpuAttr = getGPUTargetAttr(op);
+  if (!gpuAttr)
+    return failure();
+
+  StringRef chipset = gpuAttr.getArch();
+  FailureOr<amdgpu::Chipset> maybeChipset = amdgpu::Chipset::parse(chipset);
+  if (failed(maybeChipset)) {
+    op->emitOpError() << "Invalid chipset name: " << chipset;
+    return failure();
+  }
+  // Add source and target conversion types for gfx94{*} series.
+  if (maybeChipset->majorVersion == 9 && maybeChipset->minorVersion == 4) {
+    sourceTypes.insert(sourceTypes.end(), {Float8E4M3FNUZType::get(ctx),
+                                           Float8E5M2FNUZType::get(ctx)});
+    targetType = Float32Type::get(ctx);
+    return success();
+  }
+  return failure();
+}
+
+void ConvertUnsupportedFloatArithPass::runOnOperation() {
+
+  MLIRContext *context = &getContext();
+  Operation *op = getOperation();
+  SmallVector<Type> sourceTypes;
+  Type targetType;
+
+  if (failed(
+          populateSourceAndTargetType(context, op, sourceTypes, targetType))) {
+    return;
+  }
+
+  if (sourceTypes.empty()) {
+    (void)emitOptionalWarning(
+        std::nullopt,
+        "no source types specified, float emulation will do nothing");
+    return signalPassFailure();
+  }
+
+  if (llvm::is_contained(sourceTypes, targetType)) {
+    emitError(UnknownLoc::get(context),
+              "target type cannot be an unsupported source type");
+    return signalPassFailure();
+  }
+
+  TypeConverter converter;
+  arith::populateEmulateUnsupportedFloatsConversions(converter, sourceTypes,
+                                                     targetType);
+  RewritePatternSet patterns(context);
+  arith::populateEmulateUnsupportedFloatsPatterns(patterns, converter);
+  ConversionTarget target(*context);
+  arith::populateEmulateUnsupportedFloatsLegality(target, converter);
+
+  if (failed(applyPartialConversion(op, target, std::move(patterns))))
+    signalPassFailure();
```

Comment:
style nit: IREE uses braces for simple single-statement body.

```suggestion
  if (failed(applyPartialConversion(op, target, std::move(patterns)))) {
    signalPassFailure();
  }
```